src/checkers/inference/solver/strategy/GraphSolvingStrategy.java

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+       
+
+        // Merge solutions.
+        InferenceSolution solution = mergeSolution(inferenceSolutionMaps);
+       
+        //TODO: Refactor way of recording Statistics.
+        StatisticRecorder.record(StatisticKey.GRAPH_GENERATION_TIME, (graphBuildingEnd - graphBuildingStart));
+        StatisticRecorder.record(StatisticKey.GRAPH_SIZE, (long) constraintGraph.getIndependentPath().size());
+
+        return solution;
+    }
+
+    protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Constraint> constraints,
+            ProcessingEnvironment processingEnvironment) {
+        GraphBuilder graphBuilder = new GraphBuilder(slots, constraints);
+        ConstraintGraph constraintGraph = graphBuilder.buildGraph();
+        return constraintGraph;
+    }
+
+    /**
+     * Separate constraint graph, and build an underlying solver for each separated sub-graph.
+     *
+     * Sub-class may customize their own way of separating the constraint graph by overriding this method.
+     *
+     * @return a list of underlying solvers, each of them responsible for solving a separated
+     * sub-graph from the given constraint graph.
+     */
+    protected List<Solver<?>> separateGraph(ConstraintGraph constraintGraph, SolverOptions solverOptions,
+            Collection<Slot> slots, Collection<Constraint> constraints,
+            QualifierHierarchy qualHierarchy, ProcessingEnvironment processingEnvironment) {
+        Lattice lattice = new LatticeBuilder().buildLattice(qualHierarchy, slots);
+        List<Solver<?>> separatedGraphSovlers = new ArrayList<>();
+
+        for (Set<Constraint> independentConstraints : constraintGraph.getIndependentPath()) {
+            separatedGraphSovlers.add(solverFactory.createSolver(solverOptions, slots, independentConstraints,
+                    processingEnvironment, lattice, verifier));
+        }
+
+        return separatedGraphSovlers;
+    }
+
+    /**
+     * This method is called if user wants to call all underlying solvers in parallel.
+     * 
+     * @param underlyingSolvers
+     * @return A list of Map that contains solutions from all underlying solvers.
+     * @throws InterruptedException
+     * @throws ExecutionException
+     */
+    protected List<Map<Integer, AnnotationMirror>> solveInparallel(List<Solver<?>> underlyingSolvers)
+            throws InterruptedException, ExecutionException {
+
+        ExecutorService service = Executors.newFixedThreadPool(30);
+        List<Future<Map<Integer, AnnotationMirror>>> futures = new ArrayList<Future<Map<Integer, AnnotationMirror>>>();
+
+        long solvingStart = System.currentTimeMillis();
+        for (final Solver<?> underlyingSolver : underlyingSolvers) {
+            Callable<Map<Integer, AnnotationMirror>> callable = new Callable<Map<Integer, AnnotationMirror>>() {
+                @Override
+                public Map<Integer, AnnotationMirror> call() throws Exception {
```

Comment:
Open issue: consistent return type for all solve method to return `InferenceSolution`.